### DISCUSSION TITLE
Realtime closes after few seconds

### DISCUSSION THREAD

**Discussion by renushreeg** [supabase/supabase#9146] [Questions] [answered]
Hey, I'm using the realtime to get events like mouse move, transformation and others. Around 1k-3.5k updates the subscription automatically closes and errormsg is also null. Is it intended or there any issue with that?

Also attached the ws log.
![image](https://user-images.githubusercontent.com/63147175/192082550-bfead315-2271-41a7-bda7-aae9ba4abd6a.png)

Thanks for any help

**Comment by GaryAustin1** [comment]
Are you using supabase.js and do you have refresh token on?   Seems like your token might be expiring.

Also are you really updating the database with mouse type events?

**Reply by renushreeg** [reply]
No, I'm using Supabase flutter https://pub.dev/packages/supabase_flutter and yes I'm updating the database with mouse move, some image transformations value and few others.

**Reply by dshukertjr** [reply]
@renushreeg How long in minutes does it take until you see errors?

**Reply by uttapong** [reply]
This is happening to me too. I used supabase_flutter version 0.3.3 before and it's working very fine. After last week, the connection was closed after subscribing for around 5 seconds and never reconnected. I try to change to ^1.0.0-dev.8 but it problem still persists.

**Reply by dshukertjr** [reply]
@uttapong Thank you for the info. Sorry for the inconvenience. 5 seconds is very short. Does it get closed irregardless of whether you receive some realtime events or not, or does it only close is you send relatively large number of realtime events?

**Reply by dshukertjr** [reply]
@uttapong If it was working fine before and stopped working, it might a server side issue. If you could email support@supabase.com with your project ref, our realtime team could take a look and see what might be going on with your instance.

**Reply by uttapong** [reply]
Hi @dshukertjr, I received one or two realtime events before it was closed. I tried to test it by creating a new realtime database with a single record and changing its record to produce 1 event and the connection closed too. 

I'm not sure if there are any limitations of the anon key because it's the only common data shared between my production and this test.

**Reply by GaryAustin1** [reply]
One interesting change recently in realtime is there is delay after the getting subscribed until the channel actually sends updates.  This delay is typically 3 to 4ish seconds for me.  The new realtime sends a new status event to indicate it is connected (that event name has changed several times) that some of us are using to detect real connection.  I don't know if the flutter code started using that status flag or not.   Just the timing is about the same.

Edit: I would not expect to be receiving real events before that status event occurs though, so not sure how it would play here.

**Reply by renushreeg** [reply]
Yes, there is some delay 4 to 5 sec, after subscribing. But around 10 to 20 sec and 1k-3.5k updates, it closes the connection.

Also sent a mail to [support@supabase.com](mailto:support@supabase.com).

**Reply by dshukertjr** [reply]
@uttapong @renushreeg Thanks for the info. I think I am able to reproduce this with supabase_flutter. We will look into this issue. 

Let's bring the discussion to [this](https://github.com/supabase-community/supabase-flutter/issues/225) issue over on supabase-flutter repo.

**Reply by dshukertjr** [reply]
Found a bug in supabase_flutter, which is causing this issue. A new version with a fix should be released fairly soon.
https://github.com/supabase-community/realtime-dart/pull/53

**Reply by dshukertjr** [reply]
@renushreeg @uttapong 
There was a bug where if you are unauthenticated and try to listen to realtime events, it will get disconnected after 30 seconds. This issue should be resolved on [supabase_flutter v1.0.0-dev.9](https://pub.dev/packages/supabase_flutter/versions/1.0.0-dev.9). Could you try upgrading your supabase_flutter version and see if the issue is gone?

**Reply by renushreeg** [reply]
Hey, I was not able to find `on` on SupabaseQueryBuilder, is it removed?
What's the alternative?

**Reply by elhe26** [reply]
Hi @dshukertjr ,

Should we use supabase-dart instead of supabase-flutter?

**Reply by dshukertjr** [reply]
@renushreeg You can find the new realtime syntax [here](https://supabase.com/docs/reference/dart/next/subscribe) in our docs!

@elhe26 For Flutter application, it is always recommended to use supabase-flutter. supabase-dart is for non-flutter use case like server side dart, or angular-dart!

**Reply by renushreeg** [reply]
Thanks supabase team, checked with supabase_flutter 1.0.0-dev.9 and it's working properly.

**Reply by elhe26** [reply]
Hey @dshukertjr, 

I'm not getting any updates when using `stream`. Here is the implementation:

```dart

final StreamController<List<Room>> _rooms = new BehaviorSubject<List<Room>>();

String  _userId  = "test-id";

...
Stream<List<Room>> listenToRooms() {
    _supabaseClient!
        .from("rooms")
        .stream(["room_id"])
        .eq("user_id", _userId)
        .order('created_at')
        .map(
          (event) => event
              .map((element) => Room.fromJson(element))
              .toList(),
        )
        .listen((event) {
          print("Executing Room Stream");
          _rooms.add(event);
        });

    return _rooms.stream;
  }

...
```

This gets the first list and then doesn't do anything even if you modify any column in supabase dashboard.

Also, what's the difference between `on().subscribe()` vs `stream`? Is there a way to use `on().subscribe()` and filter by column?

**Reply by dshukertjr** [reply]
@elhe26 
If you are not receiving any realtime data, it's most likely because you have not enabled it in your Supabase project. You can follow the instructions in the [docs](https://supabase.com/docs/reference/dart/subscribe) to enable it!

>Realtime is disabled by default for new Projects for better database performance and security. You can turn it on by [managing replication](https://supabase.com/docs/guides/api#managing-realtime).
If you want to receive the "previous" data for updates and deletes, you will need to set REPLICA IDENTITY to FULL, like this: ALTER TABLE your_table REPLICA IDENTITY FULL;

>Also, what's the difference between on().subscribe() vs stream? Is there a way to use on().subscribe() and filter by column?

`on().subscribe()` is a method to receive realtime updates and call a callback function upon receiving them. It does not get you the initial data, so if you want the initial data, you have to fetch them using `.select()` separately. 
`.stream()` is a method that combines getting the initial data and getting realtime events and combines them as `Stream` class, which is pretty common to be used in Dart applications. You can use either one of them of your choosing, but `.stream()` typically is easier to get started with!

You can filter the result of `.on().subscribe()` by one `eq` filter. 
https://supabase.com/docs/reference/dart/subscribe#listening-to-row-level-changes

**Reply by elhe26** [reply]
@dshukertjr

 I've enabled replication and realtime broadcast for the table using the dashboard. I'm getting the initial data from the stream and nothing else after that.
 
I'm instantiating the Supabase Service class just once after the app starts and using the stream on a view but I do not see any changes when I change any value.
 
 This is the implementation with custom auth:

```dart
   ...
  Future<void> _authOnSupabase() async {
    var result = await _authenticationService.getToken();

    if (!(result is bool)) {
      var accessToken = JWT({"address": result}).sign(
        SecretKey(supabaseSecret),
      );

      _supabaseService.setAuth(accessToken);
    } 
  }
  ...
```

I've tried removing the RLS policy and seeing if there was an issue with it but I'm getting the same results.

**Reply by dshukertjr** [reply]
@elhe26 
>using the stream on a view

Supabase's realtime capability is limited to tables only at the moment, so you cannot listen to realtime updates of a view.

**Reply by elhe26** [reply]
@dshukertjr,

Sorry, I meant screen. I'm using the table as-is, not a view of the table.

**Reply by dshukertjr** [reply]
@elhe26 
Could you try if you can get realtime updates using the `.on().subscribe()` method?
https://supabase.com/docs/reference/dart/next/subscribe

If you don't get updates using `.on().subscribe()`, then there might be some misconfiguration somewhere on Supabase's end.

**Reply by elhe26** [reply]
@dshukertjr 

I'll try it right now. If this works,  Do I have to use `select` to get the initial data first?

Edit:

I saw the previous answer, sorry. We do need to call select first to retrieve initial data.

**Reply by elhe26** [reply]
@dshukertjr 

After implementing `.on().subscribe()` I got what I needed. I'll use this instead. Thanks.

**Reply by dshukertjr** [reply]
Hmm, so it seems like `.stream()` was the problem then? I wonder what the issue is.

**Reply by elhe26** [reply]
Hey @dshukertjr ,

After spending the whole night checking out why streams didn't work, I finally got the answer.

#### I'm using custom auth to perform all operations.  I'm using this [implementation](https://github.com/supabase/supabase/discussions/9146#discussioncomment-3771682) based on this [guide](https://github.com/supabase/supabase/discussions/417#discussioncomment-2002384).

This configuration is doing what it's supposed to do (inserting, updating, and selecting data) but it doesn't work well with streams. RLS policy for this configuration is not working when there's any change in the database. When I remove the policy, everything works without any issues.

**Reply by elhe26** [reply]
Hi @dshukertjr ,

As per this [comment](https://github.com/supabase-community/supabase-flutter/issues/29#issuecomment-1259357299), do we need to dispose and renew subscriptions each time we access a screen?

Something like this:

**Supabase Service**

```dart
...

StreamSubscription<List<Room>>? _roomSubscription;

...

void removeSubscription() => _roomSubscription!.cancel();

...

Stream<List<Room>> listenToRooms() {
   _roomSubscription = _supabaseClient!
        .from("rooms")
        .stream(["room_id"])
        .eq("user_id", _userId)
        .order('created_at')
        .map(
          (event) => event
              .map((element) => Room.fromJson(element))
              .toList(),
        )
        .listen((event) {
          print("Executing Room Stream");
          _rooms.add(event);
        });

    return _rooms.stream;
  }

```

The reason I'm asking is that streams stop working when you switch to another screen. When you go to that screen again, there are no changes on that view.

**Reply by elhe26** [reply]
Confirming:

The previous configuration (Dart/Flutter) allows us to renew our streams without having to use a Map (Object, Dict, etc...) to track each connection (FYI, I'm using the MVVM pattern).

**Comment by sunilguptasg** [comment]
One of the use cases is: there is a whiteboard app, where the events on the whiteboard are sent to the clients and Flutter executes the events on the client apps. So pan, zoom, etc. Also is the pen draw - which is the mouse movement recorded from pen-down to pen-up. 

These need to be recorded (and discarded later) so that the state is in sync - if a client logs in later, or disconnects and reconnects.

**Comment by uttapong** [comment]
Hi Tyler,

Thanks for you prompt support. We’re looking for the temp solutions meanwhile and hoping this issue will be fixed soon.

Kind Regards,
Uttapong Ruangrit

**Comment by renushreeg** [comment, marked as answer]
The issue has been fixed in supabase_flutter 1.0.0-dev.9.
